### PR TITLE
[qat] ConvBn2d/ConvBnReLU2d

### PR DESCRIPTION
**Pull request by jerryzh168**
Stack:
&nbsp;&nbsp;&nbsp;&nbsp;:black_circle:&nbsp; **#23024 [qat] ConvBn2d/ConvBnReLU2d**&nbsp;&nbsp;[:yellow_heart:](https://our.internmc.facebook.com/intern/diff/D16295500/)
&nbsp;&nbsp;&nbsp;&nbsp;:white_circle:&nbsp; #23022 [qat] LinearReLU module&nbsp;&nbsp;[:green_heart:](https://our.internmc.facebook.com/intern/diff/D16286800/)
&nbsp;&nbsp;&nbsp;&nbsp;:white_circle:&nbsp; #23008 [qat] ConvReLU2d module&nbsp;&nbsp;[:yellow_heart:](https://our.internmc.facebook.com/intern/diff/D16286670/)

Added _intrinsic.qat.ConvBn2d/_intrinsic.qat.ConvBnReLU2d.

Differential Revision: [D16295500](https://our.internmc.facebook.com/intern/diff/D16295500/)

### CODE REVIEW

File: torch/nn/_intrinsic/qat/modules/conv_bn_relu.py [hunk 1]
```diff
@@ -0,0 +1,193 @@
+from __future__ import absolute_import, division, print_function, unicode_literals
+from torch.nn import Conv2d as NNConv2d
+from torch.nn._intrinsic import ConvBn2d as NNConvBn2d
+from torch.nn._intrinsic import ConvBnReLU2d as NNConvBnReLU2d
+from torch.nn.modules.conv import conv2d_forwad
```
Comment (Lucaskabela) [inline torch/nn/_intrinsic/qat/modules/conv_bn_relu.py]: nit: typo, should be conv2d_forward.  See https://github.com/pytorch/pytorch/pull/22899/files#r305143145

File: torch/nn/_intrinsic/qat/modules/conv_bn_relu.py [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+            Args: `mod` a float module, either produced by torch.quantization utilities
+            or directly from user
+        """
+        assert type(mod) == cls.__FLOAT_MODULE__, ' qat.' + cls.__name__ + '.from_float only works for ' + \
+            cls.__FLOAT_MODULE__.__name__
+        if not qconfig:
+            assert hasattr(mod, 'qconfig'), 'Input float module must have qconfig defined'
+            assert hasattr(mod, 'observer'), 'Input float module must have observer attached'
+            qconfig = mod.qconfig
+        qat_convbn = cls(mod.in_channels, mod.out_channels, mod.kernel_size,
```
Comment (raghuramank100) [inline torch/nn/_intrinsic/qat/modules/conv_bn_relu.py]: Is the input to from_float a single sequential module? In that case, this code should refer to mod[0] and mod[1]. 
Also, please add a simple test that shows that numerically the forward and backward of this module matches that of a conv followed by a bn.

File: torch/nn/_intrinsic/qat/modules/conv_bn_relu.py [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+        self.weight_fake_quant = weight_fake_quant()
+
+    def reset_running_stats(self):
+        self.running_mean.zero_()
+        self.running_var.fill_(1)
+        self.num_batches_tracked.zero_()
+
+    def reset_parameters(self):
+        self.reset_running_stats()
+        if self.affine:
```
Comment (raghuramank100) [inline torch/nn/_intrinsic/qat/modules/conv_bn_relu.py]: Is self.affine defined?

File: torch/nn/_intrinsic/qat/modules/conv_bn_relu.py [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+        else:
+            conv += self.beta - self.gamma * self.running_mean / torch.sqrt(self.running_var)
+
+        return conv
+
+    def forward(self, input):
+        return self.observer(self._forward(input))
+
+    @classmethod
+    def from_float(cls, mod, qconfig):
```
Comment (raghuramank100) [inline torch/nn/_intrinsic/qat/modules/conv_bn_relu.py]: Ensure that bn module has affine=True and tracking_running_stats= True

**Review by raghuramank100 — CHANGES_REQUESTED** [review]
Please see comments, thanks

File: torch/nn/_intrinsic/qat/modules/conv_bn_relu.py (hunk 4 above)
Comment (jerryzh168) [reply torch/nn/_intrinsic/qat/modules/conv_bn_relu.py]: I think we can ensure that in conversion

File: torch/nn/_intrinsic/qat/modules/conv_bn_relu.py (hunk 3 above)
Comment (jerryzh168) [reply torch/nn/_intrinsic/qat/modules/conv_bn_relu.py]: oh, we should enforce that before conversion, I'll remove this check.

File: torch/nn/_intrinsic/qat/modules/conv_bn_relu.py (hunk 2 above)
Comment (jerryzh168) [reply torch/nn/_intrinsic/qat/modules/conv_bn_relu.py]: I think we can access all attributes of submodules in sequential?

yeah plan to add numerical tests

Comment (raghuramank100) [reply torch/nn/_intrinsic/qat/modules/conv_bn_relu.py]: What I was saying was that the from_float function should do the following (note the indices into the sequential container)
```PYTHON
    def from_float(cls, mod, qconfig):
        r"""Create a qat module from a float module or qparams_dict
            Args: `mod` a float module, either produced by torch.quantization utilities
            or directly from user
        """
        assert type(mod) == cls.__FLOAT_MODULE__, ' qat.' + cls.__name__ + '.from_float only works for ' + \
            cls.__FLOAT_MODULE__.__name__
        if not qconfig:
            assert hasattr(mod, 'qconfig'), 'Input float module must have qconfig defined'
            assert hasattr(mod, 'observer'), 'Input float module must have observer attached'
            qconfig = mod.qconfig
        qat_convbn = cls(mod[0].in_channels, mod[0].out_channels, mod[0].kernel_size,
                         stride=mod[0].stride, padding=mod[0].padding, dilation=mod[0].dilation,
                         groups=mod[0].groups, bias=mod[0].bias is not None,
                         padding_mode=mod[0].padding_mode,
                         eps=mod[1].eps, momentum=mod[1].momentum,
                         freeze_bn=False,
                         activation_fake_quant=qconfig.activation,
                         weight_fake_quant=qconfig.weight)

        qat_convbn.weight = mod[0].weight
        qat_convbn.bias = mod[0].bias
        qat_convbn.gamma = mod[1].gamma
        qat_convbn.beta = mod[1].beta
        qat_convbn.running_mean = mod[1].running_mean
        qat_convbn.running_var = mod[1].running_var
        qat_convbn.num_batches_tracked = mod[1].num_batches_tracked
        return qat_convbn
```

**Comment by jerryzh168** [comment]
why is this closed???

**Comment by jerryzh168** [comment]
@ezyang what happened?

**Comment by ezyang** [comment]
Ugh, there must be something wrong with the branch pruning script. Can you reexport it?

**Comment by jerryzh168** [comment]
yeah, I've re-exported the pr

**Pull request closed without merging**